npm_and_yarn/spec/dependabot/npm_and_yarn/yarn_error_handler_spec.rb

```diff
@@ -0,0 +1,164 @@
+# typed: false
+# frozen_string_literal: true
+
+require "spec_helper"
+require "dependabot/npm_and_yarn/file_updater/yarn_lockfile_updater"
+require "dependabot/dependency"
+require "dependabot/dependency_file"
+require "dependabot/shared_helpers"
+require "dependabot/errors"
+
+RSpec.describe Dependabot::NpmAndYarn::YarnErrorHandler do
+  subject(:error_handler) { described_class.new(dependencies: dependencies, dependency_files: dependency_files) }
+
+  let(:dependencies) { [instance_double(Dependabot::Dependency, name: "test-dependency")] }
+  let(:dependency_files) { [instance_double(Dependabot::DependencyFile, path: "path/to/yarn.lock")] }
+  let(:error) { instance_double(Dependabot::SharedHelpers::HelperSubprocessFailed, message: error_message) }
+
+  describe "#initialize" do
+    it "initializes with dependencies and dependency files" do
+      expect(error_handler.send(:dependencies)).to eq(dependencies)
+      expect(error_handler.send(:dependency_files)).to eq(dependency_files)
+    end
+  end
+
+  describe "#handle_error" do
+    context "when the error message contains a yarn error code that is mapped" do
+      let(:error_message) { "YN0002: Missing peer dependency" }
+
+      it "raises the corresponding error class" do
+        expect { error_handler.handle_error(error) }.to raise_error(Dependabot::DependencyFileNotResolvable)
+      end
+    end
+
+    context "when the error message contains a recognized pattern" do
+      let(:error_message) { "Here is a recognized error pattern: authentication token not provided" }
+
+      it "raises the corresponding error class" do
+        expect { error_handler.handle_error(error) }.to raise_error(Dependabot::PrivateSourceAuthenticationFailure)
+      end
+    end
+
+    context "when the error message contains unrecognized patterns" do
+      let(:error_message) { "This is an unrecognized pattern that should not raise an error." }
+
+      it "does not raise an error" do
+        expect { error_handler.handle_error(error) }.not_to raise_error
+      end
+    end
+
+    context "when the error message contains unrecognized yarn error codes and patterns" do
+      let(:error_message) do
+        "➤ YN0000: ┌ Resolution step\n" \
+          "➤ YN0000: ┌ Fetch step\n" \
+          "➤ YN0099: │ some-dummy-package@npm:1.0.0 can't be found\n" \
+          "➤ YN0099: │ some-dummy-package@npm:1.0.0: The remote server failed\n" \
+          "➤ YN0000: └ Completed\n" \
+          "➤ YN0000: Failed with errors in 1s 234ms"
+      end
+
+      it "does not raise an error" do
+        expect { error_handler.handle_error(error) }.not_to raise_error
```

Comment:
It seems like it should fall back to raise a default error, not eat the error.

Reply:
Actually we didn't remove current error handlings. So I am just checking in the flow if there error is a known error we can catch. Otherwise it is keep going. I am still planing to move all the error hanlings and raise error and depends to raised error catch in some situation such as time out retrying. You can check following lines.  

```
https://github.com/dependabot/dependabot-core/pull/10177/files#diff-743ecbdac7d0429bf0eb2a2f3ead58933f069a8ff82ff052a2d102e701e1887cR143
https://github.com/dependabot/dependabot-core/pull/10177/files#diff-743ecbdac7d0429bf0eb2a2f3ead58933f069a8ff82ff052a2d102e701e1887cR247
```

Reply:
But before that I would like to merge this PR since it is going to get bigger if I try to move all error handlings.